File: projects/m6/001-lotto-game/solution/mirkodallaglio/Bill.js
//All wheels
const cities = ['Bari', 'Cagliari', 'Firenze', 'Genova', 'Milano', 'Napoli', 'Palermo', 'Roma', 'Torino', 'Venezia', 'Tutte'];
//All type of bet
const types = ['Estratto','Ambetto', 'Ambo', 'Terno', 'Quaterna', 'Cinquina'];

//Bill Class
class Bill {
    _numbers;
    _city;
    _type;

    constructor (numbers, city, type) {
        this.numbers = numbers;
        this.city = city;
        this.type = type;
    };

    //getter and setter for numbers
    //verify that a number between 1 and 10 is entered, otherwise an exception is raised
    get numbers () { return this._numbers };
    set numbers (number) { 
        if (typeof number === 'number' && number <11 && number > 0) this._numbers = number
        else throw new Error (`${number} is a invalid numbers, Bill instance not created\nAccepted parameters : Number from 1 to 10`)
    };

    //getter and setter for city
    //verify that an array of cities included in const cities is entered, if all are selected only 'Tutte' will be selected
    //otherwise an exception is raised
    get city () { return this._city };
    set city (arr) {
        
        if (this.#checkInputArray(arr)) this._city = arr.filter(city => cities.includes(city));

        if (!this.#compareArray(this._city, arr)) throw new Error (`${arr} is a invalid city, Bill instance not created\nAccepted parameters : ${cities.join('-')}`)
        
        if (this.#compareArray(cities.slice(0,-1), arr)) this._city = ['Tutte'];
    };

    //getter and setter fot type
    //verify that an array of types included in const types is entered,
    //if the numbers in play are less than five, the higher plays are discarded. 
    //example: if we have only 3 numbers we cannot play Quaterna or Cinquina
    //otherwise an exception is raised
    get type () { return this._type };
    set type (arr) { 

        let removed = [];
        let message = '';
        const typesCopy = [...types];

        if (this._numbers < 5) { removed = typesCopy.splice((this._numbers !== 1) ? this._numbers+1 : this._numbers) };

        if (this.#checkInputArray(arr)) this._type = arr.filter(type => {
            if (removed.includes(type)) message = `: you cannot choose ${removed} because you are playing only ${this._numbers} numbers`;
            if (typesCopy.includes(type)) return true
        });

        if (!this.#compareArray(this._type, arr)) throw new Error (`${arr} is a invalid type${message}, Bill instance not created\nAccepted parameters : ${typesCopy.join('-')}`)

    };

    //private function that checks if the input is an array and if it contains at least one element
    // # return = true or an exception is raised
    #checkInputArray (input) {
        if (Array.isArray(input) && input.length > 0) return true;
        else throw new Error (`Input "${input}" isn't an array or it is empty, Bill instance not created`);
    };

    //private function compare 2 arrays
    // - arr1, arr2 = the 2 arrays to compare
    // # return = true or false
    #compareArray (arr1, arr2) {
        return JSON.stringify(arr1) === JSON.stringify(arr2);
    };

    //private function to generate random numbers between 1 and 90 that are never the same
    // # return = an array of numbers
    #genNumber () {
        const result = [];

        while (result.length < this._numbers) {
            const rndNum = Math.floor(Math.random() * 90) + 1;
            if (!result.includes(rndNum)) result.push(rndNum)
        };
        return result;
    };

    //private function to generates a horizontal line for the table
    // - lineWidth = number, how many characters the line should be long
    // - symbol = string, the character to use to form the line
    // # return = string
    #lineGenerator (lineWidth, symbol) {
        return `+${symbol.padStart(lineWidth-2, symbol)}+`
    };

    //private function to center a word in a space by adding spaces before and after
    // - lineWidth = number, how many characters is the space
    // - word = string, the word to write
    // # return = string
    #centerWord (lineWidth, word) {
        const wordLength = word.length;
        const space = (lineWidth - wordLength) /2;
        return word.padStart(wordLength+space, ' ').padEnd(lineWidth, ' ');
    };

    //print a ticket table with the numbers, wheels and type of bet
    // @ use #genNumber, #lineGenerator, #centerWord function
    // - ticketNumber = number, the ticket number to show in the title
    // # return = string
    print (ticketNumber) {
        const title = 'LOTTO GAME TICKET #' + ticketNumber;
        const wheel = this._city.join('  ');
        const type = this._type.join('  ');
        const rndNumber = this.#genNumber().join(' - ');
        const lineWidth = 60;

        return [title, wheel,type, rndNumber].map(el => {
            return `${this.#lineGenerator(lineWidth,'=')}\n|${this.#centerWord(lineWidth-2, el)}|`;

        }).join('\n') + '\n' + this.#lineGenerator(lineWidth,'=') + '\n\n'
        
    };

};
 
module.exports = { Bill, cities, types }

File: projects/m6/001-lotto-game/solutions/RoccoFigliamonti/bill.py
from random import sample


class Bill:

    cities = ["Bari", "Cagliari", "Firenze", "Genova", "Milano", "Napoli", "Palermo", "Roma", "Torino", "Venezia", "Tutte"]
    bet_types = ["Ambo","Terna","Quaterna","Cinquina"]

    def __init__(self, bill_num:int, city:str, bet_type:str, num_list:list):
        #assert 0 <= bill_num < 6, f"The max number of bills is 5"
        #assert bet_type.capitalize() in Bill.bet_types, f"Invalid bet type"
        #assert city.capitalize() in Bill.cities, f"Invalid city"
        self.bill_num = bill_num
        self.city = city
        self.bet_type = bet_type
        self.num_list = num_list


    def __str__(self):

        line = f"+{'-'*10}+{'-'*10}+{'-'*10}+{'-'*40}+\n"
        
        header = f"|{('TICKET N').center(10)}|{('CITY').center(10)}|{('BET').center(10)}|{('NUMBERS').center(40)}|\n"
        
        ticket = "|"+f"{self.bill_num}".center(10) + "|" + f"{self.city}".center(10) + "|" + f"{self.bet_type}".center(10) + "|" + f"{self.num_list}".center(40) + "|\n"
       
        return line + header + line + ticket + line
        


def bill_num():
    bills = int(input("How many bills do you want to generate? (type 0 to exit): "))
    if 0 < bills < 6:
        return bills
    elif bills == 0:
        print("Thanks for playing!")
        exit()
    else:
        print("The max number of bill is 5")
        bill_num()


def city_choice():
    print(Bill.cities)
    city = input("Please insert the city for this bill, choosing from the above list: ").capitalize()
    if city in Bill.cities:
        return city
    else:
        print("This is an invalid city")
        city_choice()


def bet_choice():
    print(Bill.bet_types)
    bet_type = input("Please insert the type of bet for this bill, choosing from the above list: ").capitalize()
    if bet_type in Bill.bet_types:
        return bet_type
    else:
        print("This is an invalid bet")
        bet_choice()


def ticket_number_generation():
    total_numbers_to_bet = int(input("Please insert how many numbers do you want to bet (2-10): "))
    if 1 < total_numbers_to_bet < 11:
        numbers = sample(range(1,91), total_numbers_to_bet)
        return numbers
    else:
        print("Invalid number")
        ticket_number_generation()


def ticket_generation():
    ticket_list = []
    x = bill_num()
    for i in range(1,x+1):
        city = city_choice()
        bet_type = bet_choice()
        numbers_to_bet = ticket_number_generation()
        ticket = Bill(i, city, bet_type, numbers_to_bet)
        ticket_list.append(ticket)
    return ticket_list




#main
ticket_list = ticket_generation()
for el in ticket_list:
    print(el)


